tools/pyright_tool.py

```diff
@@ -0,0 +1,658 @@
+parser.add_argument(
+    "--folder",
+    help="Folder which to test, relative to the repository root. When empty, taking the folder of this file.",
+    default="",
+)
+args = parser.parse_args()
+
+if args.dev:
+    should_generate_error_file = True
+    should_delete_error_file = False
+    print("Running in dev mode, creating the file and not deleting it")
+elif args.test:
+    should_generate_error_file = False
+    should_delete_error_file = False
+    print("Running in test mode, will reuse existing error file")
+else:
+    should_generate_error_file = True
+    should_delete_error_file = True
+
+SHOULD_GENERATE_ERROR_FILE = should_generate_error_file
+SHOULD_DELETE_ERROR_FILE = should_delete_error_file
+SHOULD_LOG = args.log
+
+if args.folder:
+    # Repository root + the wanted folder.
+    # Need to change the os folder to find all the files correctly
+    HERE = Path(__file__).resolve().parent.parent / args.folder
+    if not os.path.isdir(HERE):
+        raise RuntimeError(f"Could not find folder {args.folder} under {HERE}")
+    os.chdir(HERE)
+else:
+    # Folder of this file
+    HERE = Path(__file__).resolve().parent
+
+# TODO: move into a JSON or other config file
+# Files need to have a relative location to the folder being tested
+# Example (when checking `python` folder):
+# "tools/helloworld.py": [
+#     FileSpecificIgnore(rule="reportMissingParameterType"),
+#     FileSpecificIgnore(substring="cannot be assigned to parameter"),
+# ],
+FILE_SPECIFIC_IGNORES: FileSpecificIgnores = {}
+
+
+# Allowing for more readable ignore of common problems, with an easy-to-understand alias
+ALIASES: Dict[str, str] = {
+    "awaitable-is-generator": 'Return type of generator function must be "Generator" or "Iterable"',
+    "obscured-by-same-name": "is obscured by a declaration of the same name",
+    "int-into-enum": 'Expression of type "int.*" cannot be assigned to return type ".*"',
+}
+
+
+class PyrightTool:
+    ON_PATTERN: Final = "# pyright: on"
+    OFF_PATTERN: Final = "# pyright: off"
+    IGNORE_PATTERN: Final = "# type: ignore"
+    IGNORE_DELIMITER: Final = ";;"
+
+    original_pyright_results: PyrightResults
+    all_files_to_check: Set[str]
+    all_pyright_ignores: FileIgnores
+    pyright_off_ignores: FilePyrightOffIgnores
+    real_errors: Errors
+    unused_ignores: List[str]
+    inconsistencies: List[str] = []
+
+    def __init__(
+        self,
+        pyright_config_file: Union[str, Path],
+        *,
+        file_specific_ignores: Optional[FileSpecificIgnores] = None,
+        aliases: Optional[Dict[str, str]] = None,
+        error_file: Union[str, Path] = "temp_error_file.json",
+        should_generate_error_file: bool = True,
+        should_delete_error_file: bool = True,
+        should_log: bool = False,
+    ) -> None:
+        self.pyright_config_file = pyright_config_file
+        self.file_specific_ignores = (
+            file_specific_ignores if file_specific_ignores else {}
+        )
```

Comment:
Btw you can also use `self.file_specific_ignores = file_specific_ignores or {}`. But the if-else is fine too.